### PR TITLE
DCP 600 - Make downloading of spreadsheet async 

### PR DESCRIPTION
**Pull request by aaclan-ebi**
ebi-ait/dcp-ingest-central#600

Please check miro for screenshots: https://miro.com/app/board/uXjVOX88PlI=/

### CODE REVIEW

File: src/app/http-interceptors/http-error-interceptor.ts [hunk 1]
```diff
@@ -8,32 +15,37 @@ import {catchError, retry} from 'rxjs/operators';
 @Injectable()
 export class HttpErrorInterceptor implements HttpInterceptor {
   MAX_RETRIES = 5;
+  UNREACHABLE_SERVER = 0;
+  CONNECTION_ERROR_CODES = [
+    this.UNREACHABLE_SERVER,
```
Comment (amnonkhen) [inline src/app/http-interceptors/http-error-interceptor.ts:20]: What is the difference between 0 and 503 here?

File: src/app/shared/services/broker.service.ts [hunk 2]
```diff
@@ -33,19 +31,25 @@ export class BrokerService {
...
         {observe: 'response', responseType: 'blob'})
       .pipe(timeout(this.DOWNLOAD_SPREADSHEET_TIMEOUT),
-        map((res) => {
-          return this.getFileDataFromResponse(res);
-      }));
+        map(response => {
+          if (response.status === HttpStatusCode.Ok) {
+            return this.getFileDataFromResponse(response);
+          } else if (response.status === HttpStatusCode.Accepted) {
+            throwError(new Error("The spreadsheet is not yet available and still being generated."))
```
Comment (amnonkhen) [inline src/app/shared/services/broker.service.ts:38]: Why is this considered an error as opposed to a normal return value?

File: src/app/submission/pages/submission.component.ts [hunk 3]
```diff
@@ -277,31 +319,44 @@ export class SubmissionComponent implements OnInit, OnDestroy {
...
-      this.downloadDisabled = false;
-      this.cookieService.delete(this.DOWNLOAD_TIMEOUT_COOKIE);
-    },
-    err => {
-      if (err instanceof TimeoutError) {
+      },
+      err => {
+        const retry_message = 'Please retry later.';
+        if (err instanceof TimeoutError) {
+          this.alertService.error('', 'Spreadsheet download timed out. ' + retry_message, false, true);
```
Comment (amnonkhen) [inline src/app/submission/pages/submission.component.ts:351]: What is the empty string parameter's purpose?

File: src/app/submission/pages/submission.component.ts [hunk 4]
```diff
@@ -277,31 +319,44 @@ export class SubmissionComponent implements OnInit, OnDestroy {
...
+        } else {
+          console.error(err)
```
Comment (amnonkhen) [inline src/app/submission/pages/submission.component.ts:353]: Do we need to log the error to the console?

File: src/app/submission/pages/submission.component.ts [hunk 5]
```diff
@@ -15,10 +15,10 @@ import {AlertService} from '@shared/services/alert.service';
 import {BrokerService} from '@shared/services/broker.service';
```
Comment (amnonkhen) [inline src/app/submission/pages/submission.component.ts:15]: This file is very big.
I would bet that some of the code belongs in a Service. For example, spreadsheet generation and download stuff.

**Review by amnonkhen — CHANGES_REQUESTED** [review]

File: src/app/http-interceptors/http-error-interceptor.ts (hunk 1 above)
Comment (aaclan-ebi) [reply src/app/http-interceptors/http-error-interceptor.ts:20]: 0 is when the service is totally down and there's even no response returned. This is the status code I'm getting when I test it locally and the service is not running at all. The 503 Service Unavailable status is being returned by the service when it is not ready to handle the request yet.

File: src/app/shared/services/broker.service.ts (hunk 2 above)
Comment (aaclan-ebi) [reply src/app/shared/services/broker.service.ts:38]: The accepted status means that the file is not yet available but is already being generated. The file not found status is being returned if there is no spreadsheet being generated at all.

File: src/app/submission/pages/submission.component.ts (hunk 3 above)
Comment (aaclan-ebi) [reply src/app/submission/pages/submission.component.ts:351]: The first param is the title. I didn't think much about the error message. I'll let gabs/wrangler suggest proper error messages/titles. I just added "Success"/"Error" for now. 

Screenshot 2022-01-31 at 19 56 50

File: src/app/submission/pages/submission.component.ts (hunk 4 above)
Comment (aaclan-ebi) [reply src/app/submission/pages/submission.component.ts:353]: Might be helpful to log the full error object if there's an issue. I'm not sure what's the ideal way to report client errors.

**Review by amnonkhen — APPROVED** [review]

File: src/app/submission/pages/submission.component.ts (hunk 5 above)
Comment (aaclan-ebi) [reply src/app/submission/pages/submission.component.ts:15]: Extracted them to a new SpreadsheetTabDetailsComponent.

File: src/app/http-interceptors/http-error-interceptor.ts (hunk 1 above)
Comment (ke4) [reply src/app/http-interceptors/http-error-interceptor.ts:20]: Normally 0 return code means that everything went well, so it is a bit weird to me.

**Review by ke4 — APPROVED** [review]
Just a tiny comment that 0 return code normally means that everything went well and no error has happened. So it is strange for me to return `0` as an error code.
Otherwise it looks fine by me.

File: src/app/http-interceptors/http-error-interceptor.ts (hunk 1 above)
Comment (aaclan-ebi) [reply src/app/http-interceptors/http-error-interceptor.ts:20]: yeah but that's for bash scripts executions

>Something could go wrong on the client-side such as a network error that prevents the request from completing successfully or an exception thrown in an RxJS operator. These errors have status set to 0 and the error property contains a ProgressEvent object, whose type might provide further information.

https://angular.io/guide/http#typed-response

**Comment by aaclan-ebi** [comment]
Got approval from @amnonkhen and @ke4 ! Thanks for the reviews. Merging now.

**Pull request merged**